src/yarp.c

```diff
@@ -10105,77 +10185,14 @@ parse_heredoc_dedent(yp_parser_t *parser, yp_node_t *node, yp_heredoc_quote_t qu
             continue;
         }
 
-        // Get a reference to the string struct that is being held by the string
-        // node. This is the value we're going to actual manipulate.
-        yp_string_t *string = &(((yp_string_node_t *) node)->unescaped);
-        yp_string_ensure_owned(string);
-
-        // Now get the bounds of the existing string. We'll use this as a
-        // destination to move bytes into. We'll also use it for bounds checking
-        // since we don't require that these strings be null terminated.
-        size_t dest_length = yp_string_length(string);
-        uint8_t *source_start = (uint8_t *) string->source;
-
-        const uint8_t *source_cursor = source_start;
-        const uint8_t *source_end = source_cursor + dest_length;
-
-        // We're going to move bytes backward in the string when we get leading
-        // whitespace, so we'll maintain a pointer to the current position in the
-        // string that we're writing to.
-        uint8_t *dest_cursor = source_start;
-
-        while (source_cursor < source_end) {
-            // If we need to dedent the next element within the heredoc or the next
-            // line within the string node, then we'll do it here.
-            if (dedent_next) {
-                int trimmed_whitespace = 0;
-
-                // While we haven't reached the amount of common whitespace that we need
-                // to trim and we haven't reached the end of the string, we'll keep
-                // trimming whitespace. Trimming in this context means skipping over
-                // these bytes such that they aren't copied into the new string.
-                while ((source_cursor < source_end) && yp_char_is_inline_whitespace(*source_cursor) && trimmed_whitespace < common_whitespace) {
-                    if (*source_cursor == '\t') {
-                        trimmed_whitespace = (trimmed_whitespace / YP_TAB_WHITESPACE_SIZE + 1) * YP_TAB_WHITESPACE_SIZE;
-                        if (trimmed_whitespace > common_whitespace) break;
-                    } else {
-                        trimmed_whitespace++;
-                    }
-
-                    source_cursor++;
-                    dest_length--;
-                }
-            }
-
-            // At this point we have dedented all that we need to, so we need to find
-            // the next newline.
-            const uint8_t *breakpoint = next_newline(source_cursor, source_end - source_cursor);
-
-            if (breakpoint == NULL) {
-                // If there isn't another newline, then we can just move the rest of the
-                // string and break from the loop.
-                memmove(dest_cursor, source_cursor, (size_t) (source_end - source_cursor));
-                break;
+            yp_string_node_t *string_node = ((yp_string_node_t *) node);
+            string_node->unescaped = *(parse_heredoc_dedent_single_node(parser, &string_node->unescaped, dedent_next, common_whitespace, quote));
```

Comment:
Since you're passing a reference for the string into the function here, I don't think this assignment should be necessary.